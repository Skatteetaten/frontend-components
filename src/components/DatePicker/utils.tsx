export const DEFAULTFORMATDATE = (date: Date | null | undefined): string => {
  return date
    ? Intl.DateTimeFormat('no', {
        year: 'numeric',
        month: '2-digit',
        day: 'numeric',
      }).format(date)
    : '';
};

export const DEFAULTPARSEDATEFROMSTRING = (date: string): Date | null => {
  try {
    if (date && date.length === 10) {
      const day: number = parseInt(date.substring(0, 2));
      const month: number = parseInt(date.substring(3, 5));
      const year: number = parseInt(date.substring(6, 10));
      const formattedDate: Date = new Date(Date.UTC(year, month - 1, day));
      return formattedDate;
    }
  } catch (error) {
    console.log(error);
  }
  return null;
};

export const monthsForLocale = (
  localeName = 'no',
  monthFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { month: monthFormat })
    .format;
  const months: string[] = [];
  for (let month = 0; month < 12; month++) {
    months.push(format(new Date(Date.UTC(2020, month))));
  }
  return months;
};

export const weekdaysForLocale = (
  localeName = 'no',
  dayFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { weekday: dayFormat })
    .format;
  const weekdays: string[] = [];
  for (let day = 0; day < 7; day++) {
    weekdays.push(format(new Date(Date.UTC(2020, 7, day + 2)))); // add 2 days to let the week start on Sunday
  }
  return weekdays;
};
