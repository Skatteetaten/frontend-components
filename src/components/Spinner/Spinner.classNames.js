import { mergeStyles } from '@uifabric/merge-styles';
import { getTheme } from '@uifabric/styling';

export var getClassNames = function getClassNames(props) {
  const theme = getTheme();

  return mergeStyles([
    {
      selectors: {
        '&.ms-Spinner': {
          position: 'relative'
        },
        '.ms-Spinner-circle': {
          display: 'inline-block',
          borderRadius: '50%',
          border: '4px solid',
          borderColor:
            props.spinnerColor === 'white'
              ? theme.palette.skeColor.white
              : theme.palette.skeColor.blackAlt,
          borderTop: '4px solid transparent',
          animationDuration: '1s',
          animationIterationCount: 'linear infinite',
          animationTimingFunction: 'linear'
        }
      }
    }
  ]);
};
