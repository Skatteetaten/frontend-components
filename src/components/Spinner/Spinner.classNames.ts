import { mergeStyles } from '@uifabric/merge-styles';
import { getTheme } from '@uifabric/styling';
import { PaletteProps } from '..';
// @ts-ignore TODO
export var getClassNames = function getClassNames(props) {
  const palette = getTheme().palette as PaletteProps;

  return mergeStyles([
    {
      selectors: {
        '&.ms-Spinner': {
          position: 'relative'
        },
        '.ms-Spinner-circle': {
          display: 'inline-block',
          borderRadius: '50%',
          border: '4px solid',
          borderColor:
            props.spinnerColor === 'white'
              ? palette.skeColor.white
              : palette.skeColor.blackAlt,
          borderTop: '4px solid transparent',
          animationDuration: '1s',
          animationIterationCount: 'linear infinite',
          animationTimingFunction: 'linear'
        }
      }
    }
  ]);
};
