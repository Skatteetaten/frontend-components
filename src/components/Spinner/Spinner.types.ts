import { ISpinnerProps } from '@fluentui/react';

export interface SpinnerProps extends ISpinnerProps {
  spinnerColor?: 'white' | 'black';
}
