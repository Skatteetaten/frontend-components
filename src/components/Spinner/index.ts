export { Spinner as default } from './Spinner';
export * from './Spinner';
