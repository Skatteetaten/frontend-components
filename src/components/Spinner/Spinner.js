import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { getClassNames } from './Spinner.classNames';

import {
  Spinner as FabricSpinner,
  SpinnerSize
} from 'office-ui-fabric-react/lib-commonjs/Spinner';

/**
 * @visibleName Spinner (Innlasting)
 */
export default class Spinner extends React.PureComponent {
  static Size = SpinnerSize;
  static propTypes = {
    /** For ytterligere stiling */
    className: PropTypes.string,
    /** Farge på spinner */
    spinnerColor: PropTypes.oneOf(['white', 'black']),
    /** Størrelse på spinner */
    size: PropTypes.oneOf([
      Spinner.Size.xSmall,
      Spinner.Size.small,
      Spinner.Size.medium,
      Spinner.Size.large
    ]),
    /** Global attributt som må være unik for hele HTML dokumentet */
    id: PropTypes.string
  };
  static defaultProps = {
    size: Spinner.Size.medium,
    spinnerColor: 'black'
  };
  render() {
    const { size, className, id, spinnerColor, ...props } = this.props;

    return (
      <div id={id}>
        <FabricSpinner
          {...props}
          className={classnames(getClassNames(this.props), className)}
          spinnerColor={spinnerColor}
          size={size}
        />
      </div>
    );
  }
}
