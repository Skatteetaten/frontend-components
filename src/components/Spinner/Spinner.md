** Spinner brukes for å vise brukeren at data lastes. **

```js
import Spinner from '@skatteetaten/frontend-components/Spinner';
import Button from '@skatteetaten/frontend-components/Button';

<div style={{ textAlign: 'center' }}>
  <Button type="primary">
    {' '}
    <Spinner label="kake" spinnerColor="white" />{' '}
  </Button>
  <br />
  <br />
  <Spinner size={Spinner.Size.large} spinnerColor="black" />
</div>;
```

```js noeditor
import Accordion from '@skatteetaten/frontend-components/Accordion';
import AccordionItem from '@skatteetaten/frontend-components/Accordion/AccordionItem';

<Accordion>
  <AccordionItem
    isOpen
    toggleContent
    toggleButtonText={'Bruk'}
    stepId={'step-1-1'}
  >
    <p>
      Dersom brukeren må vente mellom ett og ti sekunder vil dette oppleves som
      en forsinkelse. I slike tilfeller er det nyttig å kunne bruke en spinner
      for å vise at systemet jobber.
    </p>
    <p>
      Spinneren kommer i svart og hvit, der svart er standardfarge. Den hvite
      spinneren kan brukes i knapper.
    </p>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Universell utforming'}
    stepId={'step-1-2'}
  >
    <ul>
      <li>
        Bruk bare spinner når du har en handling eller oppgave med innlasting.{' '}
      </li>
      <li>Vis bare én spinner om gangen.</li>
      <li>
        Spinneren kan gjerne ha en label slik at brukeren vet hva som foregår.
      </li>
      <li>
        Bruk <b>aria-label</b> for å gi elementet riktige labels.
      </li>
      <li>
        Spinneren kommer i svart og hvit, der svart er standardfarge. Den hvite
        spinneren kan brukes i knapper.
      </li>
    </ul>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Tekniske detaljer'}
    stepId={'step-1-3'}
  >
    <p>
      <a
        href="https://developer.microsoft.com/en-us/fabric#/components/spinner#Implementation"
        target="_blank"
      >
        Se flere tilgjengelige props i Fabric dokumentasjonen
      </a>
    </p>
  </AccordionItem>
</Accordion>;
```
