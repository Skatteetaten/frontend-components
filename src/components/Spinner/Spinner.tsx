import classnames from 'classnames';
import {
  ISpinnerProps,
  Spinner as FabricSpinner,
  SpinnerSize
} from 'office-ui-fabric-react/lib-commonjs/Spinner';
import * as React from 'react';
import { getClassNames } from './Spinner.classNames';

interface SpinnerProps extends ISpinnerProps {
  spinnerColor?: 'white' | 'black';
}
/**
 * @visibleName Spinner (Innlasting)
 */
export default class Spinner extends React.PureComponent<SpinnerProps, {}> {
  static Size = SpinnerSize;
  static defaultProps = {
    size: Spinner.Size.medium,
    spinnerColor: 'black'
  };
  render() {
    const { className, ...props } = this.props;
    return (
      <FabricSpinner
        {...props}
        className={classnames(getClassNames(this.props), className)}
      />
    );
  }
}
