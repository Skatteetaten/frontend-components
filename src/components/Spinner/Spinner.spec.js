import React from 'react';
import { shallow } from 'enzyme';
import { matches } from './../utils/test-utils';
import Spinner from './Spinner';

function oppsettShallow(props) {
  return shallow(<Spinner {...props} />);
}

describe('Spinner komponent', () => {
  it('matcher snapshot', () => {
    matches(<Spinner />);
  });

  it('rendrer Spinner med default props ', () => {
    const wrapper = oppsettShallow({});

    expect(wrapper.find('StyledSpinnerBase').prop('size')).toEqual(2);
    expect(wrapper.find('StyledSpinnerBase').prop('spinnerColor')).toEqual(
      'black'
    );
  });

  it('rendrer Spinner med riktige props ', () => {
    const wrapper = oppsettShallow({
      className: 'spinner-classname',
      id: 'spinner-id',
      spinnerColor: 'white',
      size: Spinner.Size.large
    });

    const StyledSpinnerBase = wrapper.find('StyledSpinnerBase');
    expect(wrapper.prop('id')).toEqual('spinner-id');
    expect(StyledSpinnerBase.prop('size')).toEqual(3);
    expect(StyledSpinnerBase.prop('className')).toContain('spinner-classname');
    expect(StyledSpinnerBase.prop('spinnerColor')).toContain('white');
  });
});
