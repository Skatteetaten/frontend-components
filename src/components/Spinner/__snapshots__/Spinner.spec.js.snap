// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner komponent matcher snapshot 1`] = `
<div
  className="ms-Spinner root-35"
>
  <div
    className="ms-Spinner-circle ms-Spinner--medium circle-36"
  />
</div>
`;
